machine/map_state.go

```diff
@@ -0,0 +1,83 @@
+package machine
+
+import (
+	"context"
+	"fmt"
+	"github.com/coinbase/step/jsonpath"
+	"github.com/coinbase/step/utils/to"
+)
+
+type MapState struct {
+	stateStr // Include Defaults
+
+	Type    *string
+	Comment *string `json:",omitempty"`
+
+	Iterator  *StateMachine
+	ItemsPath *jsonpath.Path `json:",omitempty"`
+
+	MaxConcurrency     *float64       `json:",omitempty"`
+
+	InputPath  *jsonpath.Path `json:",omitempty"`
+	OutputPath *jsonpath.Path `json:",omitempty"`
+	ResultPath *jsonpath.Path `json:",omitempty"`
+
+	Catch []*Catcher `json:",omitempty"`
+	Retry []*Retrier `json:",omitempty"`
+
+	Next *string `json:",omitempty"`
+	End  *bool   `json:",omitempty"`
+}
+//1,2,3,4
+func (s *MapState) process(ctx context.Context, input interface{}) (interface{}, *string, error) {
+	// TODO Implement
+	// FOllow what spec says will happen
```

Comment:
@grahamjenson I'm still confused about  how to complete the `process` and `execute` functions. 
I'm primary unsure about what the purpose of each should be. 

Also, you can specify both an `InputPath` and an `ItemsPath`. Should I be taking that into consideration here in my code? If so, where? Would `Execute` or `Process` make sense. 
I think I'm mostly unsure about whether or not `Execute` is fine as is right now and if `process` is where I need to update.

Reply:
the inputOuput wrapper in the `Execute` function takes into consideration the InputPath and OutputPath items. in your `process` function you should use Items path. 
I would suggest writing lots of tests to "test drive" your function to work out what is going on and make sure any assumptions are correct.